a setState pass and more logic in states

jimmyhchan opened the pull request:
Instead of the framestack in #138, this Pull Request uses state for the html attribute logic.Using the framestack we automatically received the logic that tornado bodys reset the state/stack.

Moving this html attribute code into a setState pass we have the tornado resetting happen only in the generateJS step (see util.inHtmlAttribute). Not sure if this is cleaner then #138 but this is logic may be necessary

jimmyhchan committed:
remove ctx state in favor of a setStatePass and a stateStack

the state of escapable raw and html attributes and tornado body are now
in a setState pass. The setState pass puts a state onto a node.

we introduce a stateStack which collects states from the nodes as we
build the instruction. the stateStack is an array of states we were in
as we walked the tree to build instructions.

remove dist add prepublish and npmignore

jimmyhchan commented on src/compiler/utils/FrameStack.js:
```diff
@@ -52,13 +33,20 @@ let FrameStack = function() {
+  };
+  this.popAttr = function() {
+    attrStack.leave();
+  };
```
Whoops these can be removed

smfoote commented on src/compiler/extensions/buildInstructions.js:
```diff
@@ -3,115 +3,141 @@
     frameStack.reset();
   },
-  TORNADO_PARTIAL(node, ctx, frameStack) {
+  TORNADO_PARTIAL(node, ctx, frameStack, stateStack) {
```
How many more args are we going to add to this method signature?

jimmyhchan replied:
it's limited to this file... so as many as necessary to build an instruction.

smfoote commented on src/compiler/extensions/debugger.js:
```diff
@@ -19,11 +19,11 @@ let debuggerExtension = {
         return {
           type: 'insert',
-          options: {key: node[1].key, frameStack: [inner, outer], item: node.stackItem, ctx}
+          options: {key: node[1].key, frameStack: [child, parent], item: node.stackItem, ctx}
```
no `stateStack.stack()` needed here?

jimmyhchan replied:
interesting.  you don't need it because it's not used by the codeGen step below it.

But I see your point. We need extensions to use a stable API. So we could reintroduce the Context  class, that would wrap wrap these.

smfoote commented on dist/compiler/extensions/escapableRaw.js:
```diff
@@ -4,12 +4,14 @@ var _interopRequire = function (obj) { return obj && obj.__esModule ? obj["defau
     var key = node[1].tag_info.key;
     if (escapableRawEls.indexOf(key) > -1) {
-      node[1].escapableRaw = true;
+      this.setState(node, STATES.ESCAPABLE_RAW);
```
This file should have been deleted.

smfoote commented on src/compiler/utils/Stack.js:
```diff
@@ -0,0 +1,33 @@
+    return history.length ? history[history.length - 1] : null;
+  }
+  this.current = current;
+  this.stack = function() {
```
Probably needs a new name - all the other method names are verbs. Maybe `getStack` or `getCurrentStack`

jimmyhchan replied:
i was gonna go with `peek` and `toArray`.

smfoote replied:
Those work. I think `peek` usually refers to seeing the item on top of the stack without `pop`ping it, right?

smfoote commented on src/compiler/utils/builder.js:
```diff
@@ -59,5 +59,32 @@ export default {
+
+  /**
+   * whether we are in an html_attribute given the current state
+   * NOTE: when we see a tornado body previous states are reset.
```
This doesn't make much sense to me. Can you clarify?

jimmyhchan replied:
IMO, inHTMLAttribute should be true no matter how deeply nested we are but the current `generateJS` step assumes we will create a new fragment for each tornado body. so that
`<div id="state1  {#section}state2<span id="state3 {#section2}state4{/section2}"></span>{/section"></div>`

we have 
state1 = HTML_ATTRIBUTE
state2 = TORNADO_BODY
state3=HTML_ATTRIBUTE
state4 = TORNADO_BODY
but IMO, in reality it should be

state1 = ['HTML_ATTRIBUTE']
state2 = ['HTML_ATTRIBUTE', 'TORNADO_BODY'] 
state3 = ['HTML_ATTRIBUTE', 'TORNADO_BODY', 'HTML_ATTRIBUTE'] 
state4 = ['HTML_ATTRIBUTE', 'TORNADO_BODY', 'HTML_ATTRIBUTE', 'TORNADO_BODY'] 

the calculation with lastIndex is to make this compatible with the existing generateJS but we could potentially do something smarter (throw an error that this is malformed) or optimize these to not create new fragments etc.

smfoote replied:
I like the idea of optimizing. I suppose having an html-like string in a `data-` attribute isn't actually malformed, so we should probably support this instead of throw an error.

jimmyhchan commented:
renamed `Stack.stack()` to `Stack.toArray()`, removed `dist` folder and updated gitignore and npmignore. remove unused `pushAttr` and `popAttr`. Squashed to one commit.

smfoote commented:
Nice!
